DOCSP-33454 Query Converter Content (#58)
mongodb/docs-relational-migrator: source/code-generation/query-converter/convert-queries.txt

.. _rm-convert-queries:

===============
Convert Queries
===============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

> You can convert SQL queries to MongoDB syntax with the query converter. 

Author: Updated to a slight variation of the suggested.

- .. include:: /includes/fact-query-converter-disclaimer.rst

Steps
-----

.. procedure::
   :style:  normal

   .. step:: Navigate to the query converter pane

      From the :guilabel:`Code Generation` tab, click the 
      :guilabel:`Query Converter` pane.

   .. step:: Open the SQL query converter view

      If it is your first time using the query converter in your 
      project, click :guilabel:`Paste SQL Query`.
      If you have previously converted SQL code in your project, click the
      :guilabel:`+ ADD` button on the left-hand pane next to 
      :guilabel:`Queries`.

   .. step:: Enter your SQL query

      a. Copy your SQL query to your clipboard.
      #. Under the :guilabel:`SQL Query` header, paste your SQL query
         into the text box.
   
   .. step:: Convert and test syntax

      a. Click the :guilabel:`Convert` button. Wait for your SQL code to 
         converted. The converted MongoDB syntax code displays 
         in the :guilabel:`Converted MongoDB Query` text box.
      #. Click the :icon-fa5:`copy` icon to copy the MongoDB syntax 
         to your clipboard.
      #. Test and validate the generated MongoDB syntax.

>       .. tip:: 

Reviewer: We also allow people to rename queries (just so they can be identified in the query converter), and delete them (should you no longer want to see them in the project). These are relatively minor things so they could probably be added as optional steps in this topic, rather than have whole new topics.

Author: I added a tip in step 4.1 calling out this functionality.

         .. include:: /includes/fact-query-converter-filter.rst

Examples
--------

The following table shows examples of SQL queries converted to MongoDB 

> syntax using query converter:

Reviewer: Can we add a comment:
> The syntax and structure of the converted queries will vary depending on the mapping rules used in the project.

.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - SQL Syntax
     - MongoDB Syntax

   * - 
       .. code-block:: sql
          :copyable: false

          SELECT * 
          FROM CUSTOMERS
     - 
       .. code-block:: 
          :copyable: false

          async function query(db) {
            return await db.collection('CUSTOMERS').find({
            }).toArray();
          }

   * -
      .. code-block:: sql
         :copyable: false

         SELECT CONTACTNAME, CITY 
         FROM CUSTOMERS 
         WHERE CONTACTNAME LIKE '%SMITH%'

     - 
      .. code-block:: node
         :copyable: false

         async function query(db) {

>          return await db.collection('CUSTOMERS').find({

Reviewer: All caps table/field names are common for relational databases but very rarely used in MongoDB. Can we make the examples use camelCase for the converted queries?

               CONTACTNAME: { $regex: '.*SMITH.*' }
            }, {
               projection: { CONTACTNAME: 1, CITY: 1, _id: 0 }
            }).toArray();
         }

   * -
      .. code-block:: sql
         :copyable: false

         SELECT CUSTOMERID, CITY
         FROM CUSTOMERS AS C
         JOIN ORDERS AS O
            ON C.CUSTOMERID = O.CUSTOMERID
         WHERE CONTACTNAME IN('ABI','JIM')

     -
      .. code-block:: node
         :copyable: false

         const query = async (db) => {
         return await db.collection('Customers').aggregate([
            {
               $lookup: {
               from: 'Orders',
               localField: 'CUSTOMERID',
               foreignField: 'CUSTOMERID',

>                as: 'customer_orders'

Reviewer 2: [code formatting]

The `from`, `localField`, etc fields should be indented inside of `$lookup`. Same comment applies for subfields of $match and $project.

               }
            },
            {
               $match: {
               CONTACTNAME: { $in: ['ABI', 'JIM'] }
               }
            },
            {
               $project: {
               CUSTOMERID: 1,
               CITY: 1
               }
            }
            ]).toArray();
         };

Next Steps
----------

- :ref:`rm-convert-views`
- :ref:`rm-convert-stored-procedures`

Learn More
----------

- `SQL Comparison Chart <https://www.mongodb.com/docs/manual/reference/sql-comparison/>`_

> - `SQL to Aggregation Mapping Chart <https://www.mongodb.com/docs/manual/reference/sql-aggregation-comparison/>`_

Reviewer 2: [question]

Can we use a ref for these? If no ref is available, we should use the `:manual:` docs role.

Author: Good call, updated.

.. _rm-convert-queries:

===============
Convert Queries
===============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

> You can convert SQL queries to MongoDB syntax with query converter. 

Reviewer 2: If we are updating here to say "the query converter", we should probably apply the same update throughout the page for consistency. For example, the second sentence on this page should start with "The query converter...".

Query converter considers the mapping rules and schema 
transformations defined in your project when converting your SQL code.

About this Task
---------------

.. include:: /includes/fact-query-converter-generic.rst

Before you Begin
----------------

> .. include:: /includes/fact-query-converter-disclaimer.rst

Reviewer 2: This seems like it might belong in Next Steps? I think of "Before you Begin" as something that must be done before you perform the task on the page.

Steps
-----

.. procedure::
   :style:  normal

   .. step:: Navigate to the query converter pane

>       a. From the :guilabel:`Code Generation` tab, click the 

Reviewer 2: [suggestion]

For steps that only have one substep, I'd suggest not enumerating them with `a.`. The lone substeps currently seem to provide additional context / instruction for the top-level step, so I don't really think of them as substeps.

```suggestion
      From the :guilabel:`Code Generation` tab, click the 
```

         :guilabel:`Query Converter` pane.

   .. step:: Open the SQL query converter view

      a. If it is your first time using the query converter in your 
         project, click :guilabel:`Paste SQL Query`.

>          If your project already has converted SQL code, click the 

Reviewer 2: ```suggestion
         If you have previously converted SQL code in your project, click the
```

My goal with this suggestion is to resolve the ambiguity between whether "converted" is an adjective or a verb.

         :guilabel:`+ ADD` button on the left-hand pane next to 
         :guilabel:`Queries`.

   .. step:: Enter your SQL query

      a. Copy your SQL query to your clipboard.
      #. Under the :guilabel:`SQL Query` header, paste your SQL query
         into the text box.
   
   .. step:: Convert and test syntax

      a. Click the :guilabel:`Convert` button. Wait for the AI to 
         convert your code. The converted MongoDB syntax code displays 
         in the :guilabel:`Converted MongoDB Query` text box.
      #. Click the :icon-fa5:`copy` icon to copy the MongoDB syntax 
         to your clipboard.
      #. Test and validate the generated MongoDB syntax.

      .. tip:: 

         .. include:: /includes/fact-query-converter-filter.rst

Examples
--------

The following table shows examples of SQL queries converted to MongoDB 
syntax using query converter:

.. list-table::
   :header-rows: 1
   :widths: 40 60

   * - SQL Syntax
     - MongoDB Syntax

   * - 
       .. code-block:: sql
          :copyable: false

          SELECT * 
          FROM CUSTOMERS
     - 
       .. code-block:: 
          :copyable: false

>           async function query(db) {

Reviewer 2: [question]

This looks like Node driver syntax? Do we need to specify that anywhere?

Author: @tomhollander Do you have an opinion here?

Author: Per our discussion adhoc conversions are only planned for node syntax with lightweight javascript, will mark this resolved.

         :copyable: false

         async function query(db) {
         return await db.collection('CUSTOMERS').find({
               CONTACTNAME: { $regex: '.*SMITH.*' }
            }, {
               projection: { CONTACTNAME: 1, CITY: 1, _id: 0 }
            }).toArray();
         }

   * -
      .. code-block:: sql
         :copyable: false

         SELECT CUSTOMERID, CITY
         FROM CUSTOMERS AS C
         JOIN ORDERS AS O
            ON C.CUSTOMERID = O.CUSTOMERID
         WHERE CONTACTNAME IN('ABI','JIM')

     -
      .. code-block:: node
         :copyable: false

         const query = async (db) => {
         return await db.collection('Customers').aggregate([
            {
               $lookup: {
               from: 'Orders',
               localField: 'CUSTOMERID',
               foreignField: 'CUSTOMERID',
               as: 'customer_orders'
               }
            },
            {
               $match: {
               CONTACTNAME: { $in: ['ABI', 'JIM'] }
               }
            },
            {
               $project: {
               CUSTOMERID: 1,
               CITY: 1
               }
            }
            ]).toArray();
         };

> Next Steps

Reviewer 2: [suggestion]

I'm not sure I'd consider these Next Steps. I think Next Steps are actions that have to be taken to complete a larger procedure. These links feel like they'd make more sense under Learn More.
